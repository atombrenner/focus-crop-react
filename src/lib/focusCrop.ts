export type Point = {
  x: number
  y: number
}

export type Size = {
  width: number
  height: number
}

export type Rectangle = Point & Size

export type Cropping = {
  focus: Point
  clip: Rectangle
}

export const scalePoint = ({ x, y }: Point, { width, height }: Size): Point => ({
  x: x * width,
  y: y * height,
})

export const scaleSize = (s1: Size, s2: Size): Size => ({
  width: s1.width * s2.width,
  height: s1.height * s2.height,
})

export const scaleCropping = ({ focus, clip }: Cropping, size: Size): Cropping => ({
  focus: scalePoint(focus, size),
  clip: { ...scalePoint(clip, size), ...scaleSize(clip, size) },
})

export const ratioFromString = (ratio: string) => {
  const [w, h] = ratio.split(':').map(Number)
  if (!(w && h && w > 0 && h > 0)) throw Error(`invalid ratio: ${ratio}`)
  return w / h
}

/**
 *  Calculates a rectangle with a desired ratio centered around a focus point
 *  from a source rectangle. The calculated rectangle has either the same
 *  width or the same height of the source rectangle.
 */
export const focusCrop = (ratio: number, { focus, clip }: Cropping): Rectangle => {
  // depending on the ratio keep either the width or the height of the clip rectangle
  // and calculate the other dimension from the desired ratio
  const clipRatio = clip.width / clip.height
  if (clipRatio < ratio) {
    const newHeight = clip.width / ratio
    // calculate top of rect centered around focus point y
    const newY = center(focus.y, newHeight, clip.y, clip.y + clip.height)
    return { x: clip.x, y: newY, width: clip.width, height: newHeight }
  } else {
    const newWidth = clip.height * ratio
    // calculate left of rect centered around focus point x
    const newX = center(focus.x, newWidth, clip.x, clip.x + clip.width)
    return { x: newX, y: clip.y, width: newWidth, height: clip.height }
  }
}

/** position interval with length around center and return the start of the centered interval
 *  ensures that the interval stays in [min, max] bounds
 *  if start < min then start = min
 *  if stop > max then stop = max
 *  */
const center = (center: number, length: number, min: number, max: number): number => {
  const start = center - length / 2
  return Math.min(Math.max(start, min) + length, max) - length
}
