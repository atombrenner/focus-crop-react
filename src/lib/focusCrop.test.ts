import { Cropping, focusCrop, ratioFromString } from './focusCrop'

const cropping: Cropping = {
  focus: { x: 400, y: 300 },
  clip: { x: 34, y: 37, width: 841, height: 759 },
}

const ratios = ['5:2', '16:9', '4:3', '1:1', '9:16', '90:195']

test.each(ratios)('ratio %s generates a valid rectangle', (ratio) => {
  const desiredRatio = ratioFromString(ratio)
  const rect = focusCrop(desiredRatio, cropping)
  expect(rect.width / rect.height).toBeCloseTo(desiredRatio)
  expect(rect).toMatchSnapshot()
})
