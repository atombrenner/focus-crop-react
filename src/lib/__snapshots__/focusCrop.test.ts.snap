// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ratio 1:1 generates a valid rectangle 1`] = `
Object {
  "height": 759,
  "width": 759,
  "x": 34,
  "y": 37,
}
`;

exports[`ratio 4:3 generates a valid rectangle 1`] = `
Object {
  "height": 630.75,
  "width": 841,
  "x": 34,
  "y": 37,
}
`;

exports[`ratio 5:2 generates a valid rectangle 1`] = `
Object {
  "height": 336.4,
  "width": 841,
  "x": 34,
  "y": 131.8,
}
`;

exports[`ratio 9:16 generates a valid rectangle 1`] = `
Object {
  "height": 759,
  "width": 426.9375,
  "x": 186.53125,
  "y": 37,
}
`;

exports[`ratio 16:9 generates a valid rectangle 1`] = `
Object {
  "height": 473.0625,
  "width": 841,
  "x": 34,
  "y": 63.46875,
}
`;

exports[`ratio 90:195 generates a valid rectangle 1`] = `
Object {
  "height": 759,
  "width": 350.3076923076923,
  "x": 224.84615384615387,
  "y": 37,
}
`;
